src/mos.c

```diff
@@ -459,6 +484,71 @@ int mos_cmdDIR(char * ptr) {
 	return mos_DIR(path, longListing);
 }
 
+// Assumes isxdigit(digit)
+static int xdigit_to_int(char digit) {
+	digit = toupper(digit);
+	if (digit < 'A') {
+		return digit - '0';
+	} else {
+		return digit - 55;
+	}
+}
+
+// ECHO command
+//
+int mos_cmdECHO(char *ptr) {

```

Comment:
having an echo command is cool - and I think a very useful thing to add in...  however...

I'm planning to add in system/CLI variables to the MOS CLI, in the manner that Acorn's RISC OS had (see #54 ).  as part of that I'd planned to add an `echo` that is essentially compatible with Acorn's `echo` - that allows for system variables to be interpolated into an echo output, but this uses different ways of escaping characters

Acorn's `echo` took a string that it would translate thru their `OS_GSTrans` API call.  the format for the string is documented here:
https://www.riscosopen.org/wiki/documentation/show/GS%20Translations

on an Acorn system, doing `echo \n\r` will just output those characters, and not insert a new line.  the equivalent is `echo |M|J` or `echo <13><10>`

on RISC OS as `echo` understands system variables you can do `echo Time: <Sys$Time>` and it will output a line reading something like `Time: 10:50:32`.  throw in a `<7>` in there and it'll do a VDU 7 as part of the output and beep at you.  (number support is quite sophisticated, so you can also use `<&07>` to give the value as hex, or `<16_07>` to send 7 as a base 16 value, or `<2_111>` as base 2, or `<3_21>` as base 3 if you _really_ want - or indeed any other base)

my preference in implementing system variables will be to do them in a manner similar to Acorn's - with similar names for the OS provided system variables, and providing a similar way of extracting and setting the underlying MOS sysvars.  my inclination in doing this would be to essentially follow Acorn's style of translation strings

so I think that before releasing an "official" 2.3.0 release of the C8 MOS I'd be likely to change the implementation of `echo` so that it would better match RISC OS.  that _essentially_ probably means basic GSTrans style string transformation, but excluding system variables (and not translating `"` as Acorn's `echo` seems to call OS_GSTrans with a flag set that prevents the `"` stripping).  I'd likely _not_ in the short-term implement a proper `OS_GSTrans` style API purely in the interests of getting a new release out ASAP.

NB I'm not asking you @tomm to make this change - I'll happily do it myself :grin:

Reply:
this has now been added